From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Greg Price <greg@zulip.com>
Date: Tue, 24 May 2022 18:48:07 -0700
Subject: [upstream] rnav: Fix type-parameter defaults that don't meet bounds

In the type aliases NavigationProp and NavigationHelpers.

These are all quite bogus, and Flow gives an extremely confusing error
when you try to instantiate these types and use the defaults.
TypeScript apparently just lets them slide.

Fortunately it's easy to fix: give the bound as the default, instead
of some other type that's above the bound.

TsFlower could also quite likely work around this with a hack: for
the default, emit the intersection of the original default and the
bound.  Though that's not good in some more-interesting cases, where
the default will only sometimes fail to meet the bound, and the
library author's intention is for the user to get an error in that
case.  For example, our `typesEquivalent` in src/generics.js.
---
 types/@react-navigation/core/lib/typescript/src/types.js.flow | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git types/@react-navigation/core/lib/typescript/src/types.js.flow types/@react-navigation/core/lib/typescript/src/types.js.flow
index f870405db..bfab1e949 100644
--- types/@react-navigation/core/lib/typescript/src/types.js.flow
+++ types/@react-navigation/core/lib/typescript/src/types.js.flow
@@ -199,7 +199,7 @@ type NavigationHelpersCommon<
 
 export type NavigationHelpers<
   ParamList: ParamListBase,
-  EventMap: EventMapBase = { ... },
+  EventMap: EventMapBase = EventMapBase,
 > = NavigationHelpersCommon<ParamList> &
   EventEmitter<EventMap> & {
     setParams<RouteName: $Keys<ParamList>>(
@@ -223,7 +223,7 @@ export type NavigationProp<
   NavigatorID: string | void = void,
   State: NavigationState<> = NavigationState<ParamList>,
   ScreenOptions: { ... } = { ... },
-  EventMap: EventMapBase = { ... },
+  EventMap: EventMapBase = EventMapBase,
 > = $Diff<NavigationHelpersCommon<ParamList, State>, {| getParent: mixed |}> & {
   getParent<T = NavigationProp<ParamListBase> | void>(id?: NavigatorID): T,
   setParams(params: Partial<$ElementType<ParamList, RouteName>>): void,
-- 
2.32.0

